src/prefect/orion/services/scheduler.py

```diff
@@ -134,6 +127,40 @@ def _get_select_deployments_to_schedule_query(self, db: OrionDBInterface):
         )
         return query
 
+    async def _collect_flow_runs(
+        self,
+        session: sa.orm.Session,
+        deployment_ids: List[UUID],
+        now: pendulum.DateTime,
+    ) -> List[Dict]:
+        runs_to_insert = []
+        for deployment_id in deployment_ids:
+            # guard against erroneously configured schedules
+            try:
+                runs_to_insert.extend(
+                    await self._generate_scheduled_flow_runs(
+                        session=session,
+                        deployment_id=deployment_id,
+                        start_time=now,
+                        end_time=now + self.max_scheduled_time,
+                        max_runs=self.max_runs,
+                    )
+                )
+            except Exception:
+                self.logger.exception(
+                    f"Error scheduling deployment {deployment_id!r}.",
+                )
+            finally:
+                connection = await session.connection()
+                if connection.invalidated:
+                    # If the error we handled above was a database error that caused the
+                    # transaction to rollback and become invalidated, rollback this
+                    # session, break from this loop iteration, and have the main loop
+                    # in run_once start over
+                    await session.rollback()
```

Comment:
wouldn't this rollback all work done so far including work done outside of this function?

Reply:
I don't think I'm explaining the situation in the comment well enough, I'll work on improving it, since this is so obscure.  The DB transaction has _already been rolled back_ and we're just detecting that situation with `connection.invalidated`.  Calling `session.rollback` helps to get the session back in sync with the underlying connection. That way, when we get to the end of the `with session.begin():` block in the outer function, the context manager won't attempt to `commit()` it.

Reply:
I added more words, does this make better sense?

Reply:
hmmmm okay i think im following now
would this cause issues if there's one "bad" deployment within a group that's commonly queried for together?

Reply:
(i dont think it would be any worse than it currently is, just want to guard against it)

Reply:
In both the `scheduler` and `flow-run-notifications`, we're sensitive to that problem: one bad apple spoiling the bunch, so to speak.  That's why we have the exception handling here, to guard against that case.  The only problem is that our exception handlers are also catching a small class of errors that we can't recover from unless we check the `connection.invalidated`.

I can't quite prove it, but an error like this may be an example of something we're not properly recovering from: https://github.com/PrefectHQ/nebula/issues/1157